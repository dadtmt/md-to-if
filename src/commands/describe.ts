import * as R from 'ramda'
import parseExpression, { ParsedExpression } from '../expressions'
import { TestCommandAndUpdateState, CommandUpdateState } from '.'
import { State } from '../moves'
import { SingleASTNode } from 'simple-markdown'
import { right, isRight, Either, fold, left } from 'fp-ts/lib/Either'
import { toStringIfNotString, toArrayOfStrings } from '../utils/typeCheck'

//big mistery about import from . not working
const decodeExpression: (
  parsedExpression: ParsedExpression
) => string | number = parsedExpression =>
  isRight(parsedExpression) ? parsedExpression.right : parsedExpression.left

// TODO Control errors on expressions
const getContentList: (
  state: State
) => (
  content: SingleASTNode[][],
  parsedExpressions?: (string | number)[]
) => Either<string, (string | number)[]> = state => (
  content,
  parsedExpressions = []
) => {
  const [headOfContent, ...restOfContent] = content
  if (R.isNil(headOfContent)) {
    return right(parsedExpressions)
  }
  const mayBeParsedExpression = R.pipe(
    R.head,
    R.propOr('', 'content'),
    R.split(' '),
    parseExpression(state)
  )(headOfContent)
  return fold(
    message => left(message),
    parsedExpression =>
      getContentList(state)(restOfContent, [
        ...parsedExpressions,
        parsedExpression,
      ])
  )(mayBeParsedExpression)
}

// State -> [Content] -> Object
export const getDescription: (
  state: State
) => (data: SingleASTNode[]) => Either<string, object> = state => data => {
  const headerDataLens = R.lensPath([0, 'header'])
  const cellsDataLens = R.lensPath([0, 'cells', 0])
  const mayBeHeaderContent = R.pipe(
    R.view(headerDataLens),
    getContentList(state)
  )(data)
  const mayBeCellsContent = R.pipe(
    R.view(cellsDataLens),
    getContentList(state)
  )(data)

  return fold(
    message => left(message),
    (headerContent: (string | number)[]) => {
      return fold(
        message => left(message),
        cellsContent =>
          right(R.zipObj(toArrayOfStrings(headerContent), cellsContent))
      )(mayBeCellsContent)
    }
  )(mayBeHeaderContent)
}
// R.pipe(
//   R.head,
//   R.converge(R.zipObj, [
//     R.pipe(R.prop('header'), getContentList(state)),
//     R.pipe(R.prop('cells'), R.head, getContentList(state)),
//   ])
// )

// TODO: Error "Need a key for this description"
const getDescriptionKey: (args: string[]) => string = R.pipe(
  R.head,
  R.when(R.isNil, R.always('Need a key for this description'))
)

const updateStateWithDescription: CommandUpdateState = ({
  args,
  data,
}) => state => {
  const maybeState = getDescription(state)(data)
  return fold(
    (message: string) => left(message),
    (description: object) =>
      right(R.assoc(getDescriptionKey(args), description)(state))
  )(maybeState)
}
// right(R.assoc(getDescriptionKey(args), getDescription(state)(data))(state))

const describe: TestCommandAndUpdateState = [
  R.propEq('instruction', 'describe'),
  updateStateWithDescription,
]

export default describe
