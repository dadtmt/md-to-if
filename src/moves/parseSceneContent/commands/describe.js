import * as R from 'ramda'

// [Content] -> [String]
const getContentList = R.map(R.pipe(R.head, R.prop('content')))

// [Content] -> Object
export const getDescription = R.pipe(
  R.head,
  R.converge(R.zipObj, [
    R.pipe(R.prop('header'), getContentList),
    R.pipe(R.prop('cells'), R.head, getContentList),
  ])
)

// Command -> State -> State
const describe = ({ args, data }) => R.assoc(R.head(args), getDescription(data))

export default describe
