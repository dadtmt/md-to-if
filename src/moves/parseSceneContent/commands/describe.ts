import * as R from 'ramda'
import parseExpression from './expressions'
import { Command } from '.'
import { State } from '../..'
import { Content } from '..'

type DescriptionData = {
  type: string
  header: Content[][]
  cells: Content[][]
}[]

// State -> [Content] -> [String]
const getContentList: (
  state: State
) => (content: Content[][]) => object = state =>
  R.map(R.pipe(R.head, R.prop('content'), R.split(' '), parseExpression(state)))

// State -> [Content] -> Object
export const getDescription: (
  state: State
) => (data: DescriptionData) => object = state =>
  R.pipe(
    R.head,
    R.converge(R.zipObj, [
      R.pipe(R.prop('header'), getContentList(state)),
      R.pipe(R.prop('cells'), R.head, getContentList(state)),
    ])
  )

const updateStateWithCommand: (command: Command) => (state: State) => State = ({
  args,
  data,
}) => state => R.assoc(R.head(args), getDescription(state)(data))(state)

// [Command -> Boolean, Command -> State -> State]
const describe: [
  (command: Command) => boolean,
  (command: Command) => (state: State) => State
] = [R.propEq('instruction', 'describe'), updateStateWithCommand]

export default describe
