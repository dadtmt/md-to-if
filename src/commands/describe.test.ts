import { getDescription } from './describe'

jest.mock('../expressions/rollDices')

describe('getDescription', () => {
  it('returns object description from the table', () => {
    const data = [
      {
        type: 'table',
        header: [
          [{ type: 'text', content: 'prop1' }],
          [{ type: 'text', content: 'prop2' }]
        ],
        cells: [
          [
            [{ type: 'text', content: 'val1' }],
            [{ type: 'text', content: 'val2' }]
          ]
        ]
      }
    ]

    expect(getDescription({})(data)).toMatchSnapshot()
  })
  it('handles expressions', () => {
    const data = [
      {
        type: 'table',
        header: [
          [{ type: 'text', content: 'prop1' }],
          [{ type: 'text', content: 'prop2' }],
          [{ type: 'text', content: 'numberProp' }]
        ],
        cells: [
          [
            [{ type: 'text', content: 'val1' }],
            [{ type: 'text', content: 'roll D6' }],
            [{ type: 'text', content: '12' }]
          ]
        ]
      }
    ]

    expect(getDescription({})(data)).toMatchSnapshot()
  })

  it('returns error message of the first wrong expression', () => {
    const data = [
      {
        type: 'table',
        header: [
          [{ type: 'text', content: 'prop1' }],
          [{ type: 'text', content: 'prop2' }],
          [{ type: 'text', content: 'numberProp' }]
        ],
        cells: [
          [
            [{ type: 'text', content: 'val1' }],
            [{ type: 'text', content: 'roll' }],
            [{ type: 'text', content: '12' }]
          ]
        ]
      }
    ]

    expect(getDescription({})(data)).toMatchSnapshot()
  })
})
