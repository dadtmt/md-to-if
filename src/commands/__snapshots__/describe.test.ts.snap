// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDescription handles expressions 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "numberProp": 12,
    "prop1": "val1",
    "prop2": 42,
  },
}
`;

exports[`getDescription returns error message of the first wrong expression 1`] = `
Object {
  "_tag": "Left",
  "left": "The dices are missing (ex: roll d6)",
}
`;

exports[`getDescription returns object description from the table 1`] = `
Object {
  "_tag": "Right",
  "right": Object {
    "prop1": "val1",
    "prop2": "val2",
  },
}
`;
